modules/packages/LinearAlgebra.chpl

```diff
@@ -0,0 +1,701 @@
+
+  if Arrays(1).domain.rank != 1 then compilerError("Matrix() expected 1D arrays");
+
+  const dim2 = 0..#Arrays(1).domain.dim(1).length,
+        dim1 = 0..#n;
+
+  var M: [{dim1, dim2}] eltType;
+
+  for i in dim1 do
+    M[i, ..] = Arrays(i+1)[..]: eltType;
+
+  return M;
+}
+
+
+/* Return an identity matrix over domain ``{0..#m, 0..#n}`` */
+proc eye(m, n, type eltType=real) {
+  var A: [{0..#m, 0..#n}] eltType;
+  const diagonal = min(m, n);
+  for i in 0..#diagonal do A[i, i] = 1: eltType;
+  return A;
+}
+
+
+/* Return an identity matrix over domain ``Dom`` */
+proc eye(Dom: domain(2), type eltType=real) {
+  var A: [Dom] eltType;
+  const diagonal = min(Dom.shape(1), Dom.shape(2));
+  return eye(Dom.shape(1), Dom.shape(2), eltType=eltType);
+}
+
+
+//
+// Matrix Operations
+//
+
+pragma "no doc"
+inline proc transpose(D: domain(1)) {
+  return D;
+}
+
+inline proc transpose(D: domain(2)) {
+  return {D.dim(2), D.dim(1)};
+}
+
+
+pragma "no doc"
+/* Transpose vector (no-op) */
+inline proc transpose(A: [?Dom]) where Dom.rank == 1 {
+  return A;
+}
+
+
+pragma "no doc"
+/* Syntactic sugar for transpose(Vector) */
+proc _array.T where this.domain.rank == 1 { return transpose(this); }
+
+
+/* Transpose vector or matrix.
+
+   .. note::
+
+      Since row vectors and columns vectors are indistinguishable, passing
+      a vector to this function will return that vector unchanged
+
+*/
+proc transpose(A: [?Dom] ?eltType) where Dom.rank == 2 {
+  if !isBLASType(eltType) then
+    return _transpose(A);
+  else if Dom.shape(1) == 1 then
+    return reshape(A, transpose(Dom));
+  else if Dom.shape(2) == 1 then
+    return reshape(A, transpose(Dom));
+  else {
+    var B: [Dom] eltType = eye(Dom, eltType=eltType);
+    var C: [transpose(Dom)] eltType;
+    // TODO -- use native algorithm (BLAS not necessary or memory-efficient)
+    gemm(A, B, C, 1:eltType, 0:eltType, opA=Op.T);
+    return C;
+  }
+}
+
+
+/* Transpose vector or matrix */
+proc _array.T where this.domain.rank == 2 { return transpose(this); }
+
+
+/* Transpose matrix with native implementation */
+pragma "no doc"
+private proc _transpose(A: [?Dom] ?eltType) {
+  var C: [transpose(Dom)] eltType;
+
+  // naive algorithm
+  forall (i, j) in Dom do
+    C[j, i] = A[i, j];
+
+  return C;
+}
+
+
+/* Add matrices, maintaining dimensions */
+proc matPlus(A: [?Adom] ?eltType, B: [?Bdom] eltType) {
+  if Adom.rank != Bdom.rank then compilerError("Unmatched ranks");
+  if Adom.shape != Bdom.shape then halt("Unmatched shapes");
+  var C: [Adom] eltType = A + B;
+  return C;
+}
+
+
+/* Subtract matrices, maintaining dimensions */
+proc matMinus(A: [?Adom] ?eltType, B: [?Bdom] eltType) {
+  if Adom.rank != Bdom.rank then compilerError("Unmatched ranks");
+  if Adom.shape != Bdom.shape then halt("Unmatched shapes");
+  var C: [Adom] eltType = A - B;
+  return C;
+}
+
+/*
+    Generic matrix multiplication, ``A`` and ``B`` can be a matrix, vector, or
+    scalar
```

Comment:
I suggest adding in a note here that if the user calls this with a v.M, it actually computes M^t.v, which may not be as efficient as if the user could manually swap orders.

In particular, doing a.M.b might be better implemented as dot(dot(M,b),a). This gain is only true for vectors where rows and columns are swappable.

Now that I think about it, that would be a cool helper function, since that's a common operation, and I've always found the nested dots ugly!